### DISCUSSION TITLE
Navigating the interaction between derived settings, existing setting definitions and settings inheritance

### DISCUSSION THREAD

**Discussion by DavidGregory084** [sbt/sbt#6986] [General]
Hi folks,

See also memo #6323 about derived settings.

I'd like to ask for your advice about navigating the interaction between derived settings, existing settings definitions and settings inheritance.

In sbt-tpolecat we use derived settings to ensure that when users manipulate scoped keys that relate to the plugin, they are applied only in the expected scope, so for example, if I append to `Test/tpolecatExcludeOptions` to exclude certain options from the `Test` scope, then that change should be reflected in `Test/scalacOptions`.

However, when interacting with other plugins, we also want to preserve their changes to `scalacOptions`. The defining example of this is typelevel/sbt-tpolecat#102, where we want to preserve the `-scalajs` compiler option that is applied by the Scala.js plugin for Scala 3.x.

Unfortunately when using derived settings with existing `scalacOptions`, this leads to duplication of `scalacOptions`:

```scala
    Def.derive(
      scalacOptions := {
        val previousOptions = scalacOptions.value
        val pluginOptions   = tpolecatScalacOptions.value
        val pluginExcludes  = tpolecatExcludeOptions.value
        val selectedOptions = pluginOptions.diff(pluginExcludes)
        previousOptions ++ scalacOptionsFor(scalaVersion.value, selectedOptions)
      }
    ),
```

For example, when using this in a project, we see something like this:

`Zero/scalacOptions`:
```
sbt:modeperconfiguration> show scalacOptions
[info] * -encoding
[info] * utf8
<snip>
[info] * -Ywarn-unused:privates
[info] * -Xfatal-warnings
```
`Compile/scalacOptions`
```
sbt:modeperconfiguration> show Compile/scalacOptions
[info] * -encoding
[info] * utf8
<snip>
[info] * -Ywarn-unused:privates
[info] * -Xfatal-warnings
// Duplication starts here
[info] * -encoding
[info] * utf8
<snip>
[info] * -Ywarn-unused:privates
[info] * -Xfatal-warnings
```
`Test/scalacOptions`
```
sbt:modeperconfiguration> show Test/scalacOptions
[info] * -encoding
[info] * utf8
<snip>
[info] * -Ywarn-unused:privates
[info] * -Xfatal-warnings
// Duplication #1 starts here
[info] * -encoding
[info] * utf8
<snip>
[info] * -Ywarn-unused:privates
[info] * -Xfatal-warnings
// Duplication #2 starts here
[info] * -encoding
[info] * utf8
<snip>
[info] * -Ywarn-unused:privates
```
What we can see is that for every level of settings inheritance, all of the previous options are duplicated.

This means that the compiler has a really bad time with duplicated options, and the options that the user filtered in the child scope are not actually filtered out in practice, because the parent scope's options are still applied.

It's not possible to limit the `previousOptions` to a certain scope, because we are not allowed to use scoped settings in the definition of a derived setting.

Is there a way to ensure that options applied by other plugins are preserved without dropping the derived settings behaviour?

**Comment by sjrd** [comment]
IMO all your issues come from the fact that you are using derived settings for something that they're not at all designed for.

Have you considered defining your "derived" settings through an `inConfig` filter instead?
```scala
def perConfigSettings: Seq[Setting[_]] = Def.settings(
  scalacOptions := {
    val previousOptions = scalacOptions.value
    val pluginOptions   = tpolecatScalacOptions.value
    ...
  },
)

def projectSettings: Seq[Setting[_]] = Def.settings(
  inConfig(Compile)(perConfigSettings),
  inConfig(Test)(perConfigSettings),
)
```

**Reply by DavidGregory084** [reply]
Thanks for your reply @sjrd, however it doesn't actually resolve the problem as far as I can tell - we still see duplicated settings in `Test/scalacOptions`:

```
sbt:modeperconfiguration> show Test/scalacOptions
[info] * -Xfatal-warnings
[info] * -Xlint:adapted-args
[info] * -Xlint:by-name-right-associative
[info] * -Xlint:constant
[info] * -Xlint:delayedinit-select
[info] * -Xlint:doc-detached
[info] * -Xlint:inaccessible
[info] * -Xlint:infer-any
[info] * -Xlint:missing-interpolator
[info] * -Xlint:nullary-override
[info] * -Xlint:nullary-unit
[info] * -Xlint:option-implicit
[info] * -Xlint:package-object-classes
[info] * -Xlint:poly-implicit-overload
[info] * -Xlint:private-shadow
[info] * -Xlint:stars-align
[info] * -Xlint:type-parameter-shadow
[info] * -Xlint:unsound-match
[info] * -Yno-adapted-args
[info] * -Ypartial-unification
[info] * -Ywarn-dead-code
[info] * -Ywarn-extra-implicit
[info] * -Ywarn-nullary-override
[info] * -Ywarn-nullary-unit
[info] * -Ywarn-numeric-widen
[info] * -Ywarn-unused:implicits
[info] * -Ywarn-unused:imports
[info] * -Ywarn-unused:locals
[info] * -Ywarn-unused:params
[info] * -Ywarn-unused:patvars
[info] * -Ywarn-unused:privates
[info] * -Ywarn-value-discard
[info] * -deprecation
[info] * -encoding
[info] * utf8
[info] * -feature
[info] * -language:existentials
[info] * -language:experimental.macros
[info] * -language:higherKinds
[info] * -language:implicitConversions
[info] * -unchecked
[info] * -Xlint:adapted-args
[info] * -Xlint:by-name-right-associative
[info] * -Xlint:constant
[info] * -Xlint:delayedinit-select
[info] * -Xlint:doc-detached
[info] * -Xlint:inaccessible
[info] * -Xlint:infer-any
[info] * -Xlint:missing-interpolator
[info] * -Xlint:nullary-override
[info] * -Xlint:nullary-unit
[info] * -Xlint:option-implicit
[info] * -Xlint:package-object-classes
[info] * -Xlint:poly-implicit-overload
[info] * -Xlint:private-shadow
[info] * -Xlint:stars-align
[info] * -Xlint:type-parameter-shadow
[info] * -Xlint:unsound-match
[info] * -Yno-adapted-args
[info] * -Ypartial-unification
[info] * -Ywarn-dead-code
[info] * -Ywarn-extra-implicit
[info] * -Ywarn-nullary-override
[info] * -Ywarn-nullary-unit
[info] * -Ywarn-numeric-widen
[info] * -Ywarn-unused:implicits
[info] * -Ywarn-unused:imports
[info] * -Ywarn-unused:locals
[info] * -Ywarn-unused:params
[info] * -Ywarn-unused:patvars
[info] * -Ywarn-unused:privates
[info] * -Ywarn-value-discard
[info] * -deprecation
[info] * -encoding
[info] * utf8
[info] * -feature
[info] * -language:existentials
[info] * -language:experimental.macros
[info] * -language:higherKinds
[info] * -language:implicitConversions
[info] * -unchecked
```

Initially I thought this was caused by a change to appending (`++=`) to some of the settings in the plugin rather than setting them outright (`:=`), which I needed to do to ensure that manipulating unscoped settings (`tpolecatReleaseOptions ++= somethingOrOther`) still contributed to the result, but this problem seems to occur regardless.

[Here](https://github.com/typelevel/sbt-tpolecat/pull/103)'s a PR that illustrates the problem.

> IMO all your issues come from the fact that you are using derived settings for something that they're not at all designed for.

That's most likely true, as I only know about derived settings from reading the code of sbt-typelevel and they don't appear to be clearly documented anywhere except in @eed3si9n's memo. However, it raises the question of what *are* they designed for? Should they be removed if they aren't worth documenting?

Either way, they were the only solution I could find that produces the behaviour that users expect when manipulating settings in different configurations.

**Reply by sjrd** [reply]
To avoid the duplicates, I would do something like this:
```scala
def perConfigSettings: Seq[Setting[_]] = Def.settings(
  tpolecatScalacOptions := { /* whatever you're doing today */ },
  effectivetTpolecatScalacOptions :=
    tpolecatScalacOptions.value -- (ZeroConfig / effectivetTpolecatScalacOptions ).value,
  scalacOptions := {
    val previousOptions = scalacOptions.value
    val pluginOptions   = tpolecatScalacOptions.value
    ...
  },
)

def projectSettings: Seq[Setting[_]] = Def.settings(
  tpolecatScalacOptions := { /* whatever you're doing today */ },
  effectivetTpolecatScalacOptions := tpolecatScalacOptions.value,
  inConfig(Compile)(perConfigSettings),
  inConfig(Test)(perConfigSettings),
)
```
You may need to repeat the pattern for project vs `ThisBuild` (and vs `Global`) if you also allow stuff to come from there.

**Reply by eed3si9n** [reply]
If it wasn't clear from https://github.com/sbt/sbt/discussions/6323, I find both dependency configuration and derived settings to be confusing (or cumbersome at best), and generally do not recommend using them in outside of sbt's internal code base.

If it worked correctly, a derived setting `s1` can derive its values from an existing setting `s0` automatically along all scoping axes. Since derivation doesn't work with task-scoping, this is not really suited for `scalacOptions`, I think.
